File: backend/app/api/v1/market.py
"""
Market proxy — Binance + Yahoo Finance via backend
- CORS koruması: frontend doğrudan dış API'ye gitmez, backend üzerinden geçer
- 3.5s in-memory cache per symbol (spec 3)
- Auth gerekli: get_current_user (sadece giriş yapmış trader)
"""
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
import time
import asyncio
import httpx
from typing import Dict, Any

from app.core.database import get_db
from app.core.security import get_current_user
from app.models.user import User

router = APIRouter(prefix="/market", tags=["market"])

# Cache: symbol -> (price, source, ts)
_CACHE: Dict[str, tuple[float, str, float]] = {}
_TTL = 3.5
_LOCK = asyncio.Lock()

def _norm(s: str) -> str:
    return s.strip().upper().replace("/", "").replace("-", "")

def _is_crypto(s: str) -> bool:
    n = _norm(s)
    return bool(
        n.endswith(("USDT", "USDC", "BUSD", "BTC", "ETH", "BNB", "TRY", "EUR"))
        or n in {"BTCUSDT","ETHUSDT","SOLUSDT","BNBUSDT","XRPUSDT","ADAUSDT","AVAXUSDT","DOTUSDT","DOGEUSDT","MATICUSDT"}
    )

async def _fetch_binance(symbol: str, client: httpx.AsyncClient) -> float | None:
    n = _norm(symbol)
    if not _is_crypto(n):
        return None
    urls = [
        f"https://data-api.binance.vision/api/v3/ticker/price?symbol={n}",
        f"https://api.binance.com/api/v3/ticker/price?symbol={n}",
        f"https://api1.binance.com/api/v3/ticker/price?symbol={n}",
    ]
    for url in urls:
        try:
            r = await client.get(url, timeout=4.0)
            if r.status_code != 200:
                continue
            j = r.json()
            price_str = j.get("price") if isinstance(j, dict) else None
            if price_str:
                p = float(price_str)
                if p > 0:
                    return p
        except Exception:
            continue
    return None

async def _fetch_yahoo(symbol: str, client: httpx.AsyncClient) -> float | None:
    n = _norm(symbol)
    candidates: list[str] = [n]
    if n.isalpha() and 3 <= len(n) <= 6 and "." not in n and not _is_crypto(n):
        candidates.append(f"{n}.IS")
    if symbol.strip().upper() != n:
        candidates.insert(0, symbol.strip().upper())
    for sym in candidates:
        try:
            url = f"https://query1.finance.yahoo.com/v8/finance/chart/{sym}?interval=1m&range=1d"
            r = await client.get(url, timeout=4.0, headers={"User-Agent": "Mozilla/5.0"})
            if r.status_code != 200:
                continue
            j = r.json()
            result = (j.get("chart") or {}).get("result", [{}])[0] if isinstance(j, dict) else None
            if not result:
                continue
            meta_price = result.get("meta", {}).get("regularMarketPrice")
            if isinstance(meta_price, (int, float)) and meta_price > 0:
                return float(meta_price)
            quote = result.get("indicators", {}).get("quote", [{}])[0].get("close") if result.get("indicators") else None
            if isinstance(quote, list):
                for v in reversed(quote):
                    if isinstance(v, (int, float)) and v > 0:
                        return float(v)
        except Exception:
            continue
    return None

@router.get("/prices")
async def get_prices(
    symbols: str = Query(..., description="CSV örn: BTCUSDT,ETHUSDT,THYAO veya tek sembol"),
    db: AsyncSession = Depends(get_db),
    current: User = Depends(get_current_user),
) -> Dict[str, Any]:
    # symbols can be "BTCUSDT,ETHUSDT" or "BTCUSDT"
    raw_list = [s.strip() for s in symbols.split(",") if s.strip()]
    # also support repeated query ?symbols=BTCUSDT&symbols=ETHUSDT is not parsed by this, but we handle CSV
    # normalize dedup preserve order
    seen = set()
    uniq: list[str] = []
    for s in raw_list:
        n = _norm(s)
        if n and n not in seen:
            seen.add(n)
            uniq.append(s.strip())
    if not uniq:
        raise HTTPException(400, "symbols gerekli")
    if len(uniq) > 20:
        raise HTTPException(400, "En fazla 20 sembol")

    now = time.time()
    result: Dict[str, Any] = {}
    to_fetch: list[str] = []

    async with _LOCK:
        for s in uniq:
            n = _norm(s)
            cached = _CACHE.get(n)
            if cached and (now - cached[2]) < _TTL:
                price, source, ts = cached
                result[n] = {"price": price, "source": source, "ts": int(ts * 1000)}
            else:
                to_fetch.append(s)

    if to_fetch:
        async with httpx.AsyncClient() as client:
            # fetch concurrently
            async def fetch_one(sym: str) -> tuple[str, float | None, str | None]:
                n = _norm(sym)
                p = None
                src = None
                if _is_crypto(n):
                    p = await _fetch_binance(sym, client)
                    if p is not None:
                        src = "binance"
                if p is None:
                    p2 = await _fetch_yahoo(sym, client)
                    if p2 is not None:
                        p = p2
                        src = "yahoo"
                if p is None and _is_crypto(n):
                    # fallback yahoo for crypto CORS
                    alt = n.replace("USDT", "-USD").replace("TRY", "-TRY")
                    p3 = await _fetch_yahoo(alt, client)
                    if p3 is not None:
                        p = p3
                        src = "yahoo"
                return n, p, src

            fetched = await asyncio.gather(*(fetch_one(s) for s in to_fetch))
            async with _LOCK:
                for n, p, src in fetched:
                    if p is not None and src:
                        ts = time.time()
                        _CACHE[n] = (p, src, ts)
                        result[n] = {"price": p, "source": src, "ts": int(ts * 1000)}
                    else:
                        # not found, keep missing (frontend fallback)
                        pass

    # Ensure we return only requested symbols that were found; missing ones omitted
    return result

@router.get("/price/{symbol}")
async def get_single_price(
    symbol: str,
    db: AsyncSession = Depends(get_db),
    current: User = Depends(get_current_user),
) -> Dict[str, Any]:
    res = await get_prices(symbols=symbol, db=db, current=current)
    n = _norm(symbol)
    if n not in res:
        raise HTTPException(404, f"Fiyat bulunamadı: {symbol}")
    return {n: res[n]}

File: frontend/lib/market.ts
"use client"
// Market live price service — Binance (crypto) + Yahoo Finance (stocks/ETF) free public APIs
// Spec 2.2: favori ürünlerin canlı fiyatını çek, Trades form entry_price otomatik doldur.
// CORS-friendly: Binance data-api.binance.vision supports CORS; Yahoo may block → graceful fallback.

export type PriceResult = { price: number; source: "binance" | "yahoo" | "cache"; ts: number }

const CACHE = new Map<string, PriceResult>()
const CACHE_TTL_MS = 3500

function normSymbol(s: string): string {
  return s.trim().toUpperCase().replace("/", "").replace("-", "")
}

function isCryptoSymbol(s: string): boolean {
  const n = normSymbol(s)
  // Binance quote suffixes
  return /(USDT|USDC|BUSD|BTC|ETH|BNB|TRY|EUR)$/.test(n) || ["BTCUSDT","ETHUSDT","SOLUSDT","BNBUSDT","XRPUSDT","ADAUSDT","AVAXUSDT","DOTUSDT","DOGEUSDT","MATICUSDT"].includes(n)
}

async function fetchBinance(symbol: string): Promise<number | null> {
  const n = normSymbol(symbol)
  if (!isCryptoSymbol(n)) return null
  const urls = [
    `https://data-api.binance.vision/api/v3/ticker/price?symbol=${n}`,
    `https://api.binance.com/api/v3/ticker/price?symbol=${n}`,
    `https://api1.binance.com/api/v3/ticker/price?symbol=${n}`,
  ]
  for (const url of urls) {
    try {
      const r = await fetch(url, { cache: "no-store" })
      if (!r.ok) continue
      const j = await r.json()
      const p = parseFloat(j.price || j[0]?.price)
      if (Number.isFinite(p) && p > 0) return p
    } catch {}
  }
  return null
}

async function fetchYahoo(symbol: string): Promise<number | null> {
  // Yahoo suffix mapping for BIST: THYAO -> THYAO.IS
  const n = normSymbol(symbol)
  const candidates = [n]
  // if seems BIST without .IS, try with .IS
  if (/^[A-Z]{3,6}$/.test(n) && !n.includes(".") && !isCryptoSymbol(n)) {
    candidates.push(`${n}.IS`)
  }
  // also try raw symbol as-is (user may include .IS)
  if (symbol !== n) candidates.unshift(symbol.trim().toUpperCase())
  for (const sym of candidates) {
    try {
      const url = `https://query1.finance.yahoo.com/v8/finance/chart/${encodeURIComponent(sym)}?interval=1m&range=1d`
      const r = await fetch(url, { cache: "no-store" })
      if (!r.ok) continue
      const j = await r.json()
      const result = j.chart?.result?.[0]
      const metaPrice = result?.meta?.regularMarketPrice
      if (Number.isFinite(metaPrice) && metaPrice > 0) return metaPrice
      const quote = result?.indicators?.quote?.[0]?.close
      if (Array.isArray(quote)) {
        for (let i = quote.length - 1; i >= 0; i--) {
          if (Number.isFinite(quote[i]) && quote[i] > 0) return quote[i]
        }
      }
    } catch {}
  }
  return null
}

export async function getLivePrice(rawSymbol: string): Promise<PriceResult | null> {
  const sym = normSymbol(rawSymbol)
  if (!sym) return null
  const cached = CACHE.get(sym)
  if (cached && Date.now() - cached.ts < CACHE_TTL_MS) return cached
  let price: number | null = null
  let source: PriceResult["source"] = "cache"
  // crypto first → Binance
  if (isCryptoSymbol(sym)) {
    price = await fetchBinance(sym)
    if (price != null) source = "binance"
    else {
      // fallback Yahoo for crypto? skip
    }
  }
  // if not crypto or binance failed, try Yahoo
  if (price == null) {
    price = await fetchYahoo(rawSymbol)
    if (price != null) source = "yahoo"
  }
  // also for crypto try Yahoo as second fallback (e.g. CORS blocked)
  if (price == null && isCryptoSymbol(sym)) {
    const y = await fetchYahoo(sym.replace("USDT","-USD").replace("TRY","-TRY"))
    if (y != null) { price = y; source = "yahoo" }
  }
  if (price == null || !Number.isFinite(price)) return null
  const res: PriceResult = { price, source, ts: Date.now() }
  CACHE.set(sym, res)
  return res
}

export async function getLivePrices(symbols: string[]): Promise<Record<string, PriceResult>> {
  const uniq = [...new Set(symbols.map(normSymbol).filter(Boolean))]
  const out: Record<string, PriceResult> = {}
  await Promise.all(uniq.map(async (s) => {
    const r = await getLivePrice(s)
    if (r) out[s] = r
  }))
  return out
}

export function formatPrice(price: number | null | undefined, symbol: string): string {
  if (price == null || !Number.isFinite(price as number)) return "—"
  const p = price as number
  if (p >= 1000) return p.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })
  if (p >= 1) return p.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 4 })
  return p.toLocaleString("en-US", { minimumFractionDigits: 4, maximumFractionDigits: 8 })
}

// helper to compute floating R for OPEN trades given live price map
export function computeFloatingR(trade: any, livePrice: number | null): number | null {
  if (!trade || trade.status !== "OPEN") return null
  if (livePrice == null || !Number.isFinite(livePrice)) return null
  const entry = trade.entry_price
  const sl = trade.stop_loss
  const size = trade.position_size
  if (!entry || !size) return null
  const dir = trade.direction
  const pnlCash = dir === "LONG" ? (livePrice - entry) * size : (entry - livePrice) * size
  const fees = trade.commission_fees || 0
  const netCash = pnlCash - fees
  const riskPerUnit = sl ? Math.abs(entry - sl) : entry * 0.02
  if (!riskPerUnit || riskPerUnit === 0) return null
  const pnlR = netCash / (riskPerUnit * size)
  return Math.round(pnlR * 1000) / 1000
}

export function computeFloatingCash(trade: any, livePrice: number | null): number | null {
  if (!trade || trade.status !== "OPEN") return null
  if (livePrice == null) return null
  const entry = trade.entry_price
  const size = trade.position_size
  const dir = trade.direction
  const pnlCash = dir === "LONG" ? (livePrice - entry) * size : (entry - livePrice) * size
  const fees = trade.commission_fees || 0
  return Math.round((pnlCash - fees) * 100) / 100
}
